File: Python/HackerRank/One_Week_Preparation_Kit/Day_4/mock_test.py
#
# Complete the 'truckTour' function below.
#
# The function is expected to return an INTEGER.
# The function accepts 2D_INTEGER_ARRAY petrolpumps as parameter.
#

def truckTour(petrolpumps):
    # Write your code here
    # [[1, 5], [10, 3], [3, 4]] amount of petrol, distance / 1 
    result = 0
    tank   = 0

    for i in range(len(petrolpumps)):
        tank += petrolpumps[i][0]

        if tank < petrolpumps[i][1]:
            tank   = 0
            result = i + 1 if i < len(petrolpumps) else i
        else:
            tank -= petrolpumps[i][1]

    return result

File: Javascript/HackerRank/1 Week Preparation Kit/Day 4/mock_test.js
// https://www.hackerrank.com/interview/preparation-kits/one-week-preparation-kit/one-week-day-four/challenges

'use strict';

const fs = require('fs');

process.stdin.resume();
process.stdin.setEncoding('utf-8');

let inputString = '';
let currentLine = 0;

process.stdin.on('data', function(inputStdin) {
    inputString += inputStdin;
});

process.stdin.on('end', function() {
    inputString = inputString.split('\n');

    main();
});

function readLine() {
    return inputString[currentLine++];
}



/*
 * Complete the 'truckTour' function below.
 *
 * The function is expected to return an INTEGER.
 * The function accepts 2D_INTEGER_ARRAY petrolpumps as parameter.
 */

function truckTour(petrolpumps) {
    let result = 0;
    let tank   = 0;

    for (let i = 0; i < petrolpumps.length; i++) {
        tank += petrolpumps[i][0]

        if (tank < petrolpumps[i][1]) {
            tank   = 0;
            result = i < petrolpumps.length ? i + 1 : i;
        } else {
            tank -= petrolpumps[i][1];
        }
    }

    return result;
}

function main() {
    const ws = fs.createWriteStream(process.env.OUTPUT_PATH);

    const n = parseInt(readLine().trim(), 10);

    let petrolpumps = Array(n);

    for (let i = 0; i < n; i++) {
        petrolpumps[i] = readLine().replace(/\s+$/g, '').split(' ').map(petrolpumpsTemp => parseInt(petrolpumpsTemp, 10));
    }

    const result = truckTour(petrolpumps);

    ws.write(result + '\n');

    ws.end();
}
